src/app/modules/react-migration/react-migration.controller.ts

```diff
@@ -49,24 +132,45 @@ const serveFormReact: ControllerHandler = (_req, res) => {
       reactFrontendPath,
     },
   })
+  const formId = req.params.formId
+  const createMetatagsResult = await createMetatags({
+    formId,
+  })
+
+  let reactHtmlWithMetaTags
+  // Failed to create metatags.
+  if (createMetatagsResult.isErr()) {
+    logger.error({
+      message: 'Error fetching metatags',
+      meta: {
+        action: 'handleRedirect',
+        ...createReqMeta(req),
+      },
+      error: createMetatagsResult.error,
+    })
+    reactHtmlWithMetaTags = replaceWithMetaTags({
+      title: 'FormSG',
+      description: '',
+      image: 'og-img-metatag-publicform.png',
+    })
+  } else {
+    const { title, description } = createMetatagsResult.value
+    reactHtmlWithMetaTags = replaceWithMetaTags({
+      title: title,
+      description: description ?? '',
+      image: 'og-img-metatag-publicform.png',
+    })
+  }
+
   return (
     res
       // Prevent index.html from being cached by browsers.
       .setHeader('Cache-Control', 'no-cache')
-      .sendFile(path.join(reactFrontendPath, 'index.html'))
+      .send(reactHtmlWithMetaTags)
```

Comment:
This file still contains 3 calls to `replaceWithMetaTags()`, and 2 calls `.setHeader('Cache-Control', 'no-cache')` 🤔 . I reckon the flow can be improved with a processing pipeline where the different conditional flows only collect the meta tag data, but there's a single shared final action to set the header and send the body.